### PR TITLE
Synchronous dashboard refreshes

### LINKED ISSUE #4752: Dashboard reloads don't spark joy
## Bug description

Not really a bug, but annoying UX-wise.

When clicking "reload" to reload a dashboard, the entire screen goes blank with a spinner, that spins until all dashboard items reload in the background. Only then is anything shown to the user.

## Expected behavior

Clicking reload, the screen stays as is and all individual dashboard items get a spinner. Dashboard items that finish earlier can be interacted with immediately.

## How to reproduce

1. Load a heavier dashboard such as this [internal one](https://app.posthog.com/dashboard/2256)
2. Click "reload"
3. Wait
4. Wait
5. Wait
6. Wait
7. Wait
8. Open the devtools console to check for errors
9. See none and give up
10. Oh it loads now

## Environment

- [ ] PostHog Cloud
- [ ] self-hosted PostHog, version/commit: _please provide_

## Additional context

- Most likely introduced by https://github.com/PostHog/posthog/pull/4598
- We [should use progress bars](https://www.nngroup.com/articles/progress-indicators/) or otherwise indicate progress for any loading screen that takes more than 10 seconds. Use count if time can't be estimated. E.g. "Refreshing dashboards items: 4 / 20 [spinner]"

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Closes #4752.
Historically related to #4598.

- [x] Make dashboard item calls synchronous
- [x] Add loading indicator
- [x] Add progress indicator and change last refreshed copy
- [x] Trigger refresh when auto refresh is turned on
- [x] Add custom auto reload dropdown (default 5 min)
- [x] Fix existing bug where reloading dashboard with paths graph was redirecting to paths insight graph page.

### Before

Screen Shot 2021-08-02 at 1 44 12 PM

https://user-images.githubusercontent.com/13460330/127921516-48f2aaf2-84ea-4ce3-937b-04eefe8882a3.mov

### After

Screen Shot 2021-08-02 at 1 44 30 PM

Screen Shot 2021-08-02 at 3 07 13 PM

https://user-images.githubusercontent.com/13460330/127929848-4354c825-7781-4763-a3c5-b83ea37aecd4.mov

Borrowing a dropdown pattern I've seen being used in the AWS console.

Screen Shot 2021-08-02 at 1 48 05 PM

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Comment by clarkus** [comment]
This looks like a great start. Loading states are the only thing that stand out to me as something we could talk about improving more. Even though this is a manual refresh, depending on the time it takes to reload, we might see an improvement to "perceived performance" if we use skeleton loading states. This can be tricky depending on the insight type and the options applied, but the suggestion of structure and data can make it feel like each chart is loading faster. Another thing to think about is when reload times are fast. Showing the spinner, then quickly swapping in the real data can be fine, but it would be a more seamless transition from skeleton state to data state.

**Review by paolodamico — COMMENTED** [review]
Some early feedback:
- Concept looks super solid, love it! +1 to @clarkus's suggestion of using skeletons instead of the spinner, makes the experience feel faster/more intuitive.
- I would add a general progress indicator, it could be a simple spinner with a "Loaded M of N dashboards" but maybe @clarkus has a better suggestion.
- UX-wise we may want to either hide or change the copy of the "Last refresh" as it's not updated until all dashboards get updated.

Do tag me again for another review when it makes sense 😉

**Comment by alexkim205** [comment]
Got it! I agree skeleton states would be a good idea. Talked offline with @clarkus that empty skeleton states isn't a blocker and can be implemented into our dashboards retroactively.

+1 I think having some progress indicator like the one we have in our plugins page could be informative for our heavy dashboard users.

**Review by clarkus — APPROVED** [review]
These changes look good to me Ux-wise. I'd double check that the 10s auto-refresh will work well for us. If our chart load times are approaching multiple seconds then it could result in a lot of flashing / reloading and could interrupt being able to view a chart. I see that the user can always increase the increment, but maybe 10s is too disruptive? I also see that manual refresh times aren't considered with auto refresh. For example, I can manually click refresh towards the end of the 10s timeframe, and I'll see two refreshes chained together. tl;dr - the timing on this could be tricky and disruptive for some users. Maybe a larger minimum refresh increment and balancing against manual refreshes would improve the experience? Oh we should also use a triangle down / dropdown icon instead of the ellipsis icon here.

**Comment by alexkim205** [comment]
Thanks for the review! I agree, 10s would be a bit too short so I can remove that one for now. 

I think there's a way to reset timeouts whenever a user manually refreshes. I think including a manual refresh with the interval makes the most sense.

File: frontend/src/types.ts [hunk 1]
```diff
@@ -887,6 +893,11 @@ export enum DashboardItemMode {
     Edit = 'edit',
 }
 
+export interface DashboardAutoRefresh {
```
Comment (paolodamico) [inline frontend/src/types.ts:899]: This seems like a very specific type, perhaps should be added directly on `dashboardLogic`? Further, it's not actually used at all right now.

File: frontend/src/types.ts [hunk 2]
```diff
@@ -478,6 +478,12 @@ export interface DashboardType {
     tags: string[]
 }
 
+export interface DashboardRefreshStatusType {
```
Comment (paolodamico) [inline frontend/src/types.ts:484]: This type isn't used anywhere

File: frontend/src/scenes/paths/pathsLogic.ts [hunk 3]
```diff
@@ -182,12 +182,16 @@ export const pathsLogic = kea<pathsLogicType<PathNode, PathResult>>({
             (featureFlags, loaded) => !featureFlags[FEATURE_FLAGS.TAXONOMIC_PROPERTY_FILTER] && !loaded,
         ],
     },
-    actionToUrl: ({ values }) => ({
+    actionToUrl: ({ values, props }) => ({
         setProperties: () => {
-            return ['/insights', values.propertiesForUrl, undefined, { replace: true }]
+            if (!props.dashboardItemId) {
```
Comment (paolodamico) [inline frontend/src/scenes/paths/pathsLogic.ts:187]: Great catch!

File: frontend/src/scenes/dashboard/dashboardLogic.js [hunk 4]
```diff
@@ -133,15 +149,23 @@ export const dashboardLogic = kea({
...
         ],
         autoRefresh: [
-            false,
             {
-                setAutoRefresh: (_, { enabled }) => enabled,
+                interval: AUTO_REFRESH_INITIAL_INTERVAL_SECONDS,
+                enabled: false,
+            },
+            {
+                setAutoRefresh: (state, { enabled, interval }) => ({ ...state, enabled, interval }),
```
Comment (paolodamico) [inline frontend/src/scenes/dashboard/dashboardLogic.js:157]: Nit: Is destructuring necessary? We will overwrite both params anyways

**Review by paolodamico — APPROVED** [review]
Added a few comments inline. Some additional issues I found:
1. For some reason, sometimes when clicking the "..." button, the refresh text will change (the last refresh time will get updated) which is super confusing because you think it changed because you clicked on that button.
2. If I switch between the options continuously, this error pops up.
    
3. I would try to keep the width of the button when the text changes because dashboards are loading to avoid you having to spend time figuring out what's going on.
4. The dropdown should have a lower z-index than the top navigation

File: frontend/src/types.ts (hunk 2 above)
Comment (alexkim205) [reply frontend/src/types.ts:484]: Nice catch, forgot to remove this. I wish `dashboardLogic.js` was typescriptified, which I'll probably tackle in a future PR.

**Comment by alexkim205** [comment]
Thank you all! Pushed some changes that address

- [X] Reset interval upon manual refresh
- [X] change dropdown icon to caret down icon
- [X] fix breakpoint errors
- [X] z-indexing error
- [X] consistent widths


https://user-images.githubusercontent.com/13460330/128079086-111c19f5-a348-4349-a194-58799cc1ea50.mov



As for 

> For some reason, sometimes when clicking the "..." button, the refresh text will change (the last refresh time will get updated) which is super confusing because you think it changed because you clicked on that button.

I don't know if I have a good workaround for this. Perhaps consistent widths should help with this?

**Comment by paolodamico** [comment]
Let me know if you'd like me to take another pass. One non-blocking thing I forgot to mention is using skeletons instead of spinners for the loading state of each dashboard item.

**Comment by macobo** [comment]
Hi @alexkim205 what sort of feedback would be useful from my side?

**Comment by clarkus** [comment]
I am going to produce loading skeleton states for each chart type in https://github.com/PostHog/posthog/issues/5411. The change in loading style is still a huge improvement over what we had before. I think it'd be good to queue up implementing skeleton states once I have those designs done. I can re-prioritize those if you think they're critical, @paolodamico

**Comment by alexkim205** [comment]
no worries, I think I got all the feedback I was looking for

**Comment by macobo** [comment]
There are code-level issues which I didn't point out due to being unsure due to a lot of open discussion around the _what/why_:

- This PR now introduced dead code in the backend around dashboard refreshes. We should avoid that.
- Code is not race-safe. If user navigates away from dashboards before data loads, errors occur but are ignored due to the try-catch catching all errors. Same elsewhere - if something is asynchronous, it needs to handle user navigating away/logics getting unmounted.
- Reloading all dashboard items at once is dangerous. It seems much more sensible to set a limit (e.g. 8) to fetch at the same time instead of overloading the backend at once
- It's not clear if/how error states are handled. If a particular item fails to load, how is that presented.
- Code nit: All other z-indexes are defined as consts in a single file. Adding a random class violates that.
- Code nit: code in `refreshAllDashboardItems` seems generally crufty. E.g. the `forEach` serves no real purpose, not sure about usage of breakpoint(), over-eager try-catch, etc.

To make reviews clearer in the future, perhaps worth following this flow:
1. Either draft-pr or feature-flag the new flow. Keep one review focused on a single thing - either the flow itself or the code. Having multiple people handle both is confusing. You can (should!) always remove the flag later.
2. If PR is multi-stage, communicate clearly expectations - e.g. "you're here for code quality, you're here for feedback on the flow". Otherwise important things get missed. Checklists help!

**Comment by alexkim205** [comment]
Hey Karl, although I wish this had come up during this PR review, these are still pretty important quality issues that I'd like to fix in a second iteration, and definitely appreciate the feedback.

I think the dead code you're referring to is the `api/dashboard` endpoint which we no longer use. The new code hits `api/dashboard_item` for each dashboard item refresh, but I think it makes sense that we try to load this data in chunks (i.e., given 10 total items, load 4, then the next 4, then the last 2). This batched operation will relieve some burden from the backend and can be implemented by adding `limit` and `offset` parameters to the `api/dashboard` endpoint. 4 seems like a practical denomination given that most dashboards aren't more than 10 items or so, but let me know if you disagree.

I think this PR deals w dashboard flows from a UX flow perspective, so I'll take another pass at this w/ an emphasis on code quality (+ above backend modifications, error handling) and make sure to CC you!

**Comment by macobo** [comment]
Sounds perfect, thanks Alex. The main deal is that these issues get dealt in an eventually consistent way :)

**Pull request merged**